### PR TITLE
[optimization] open gcc strict-aliasing optimization

### PR DESCRIPTION
**Pull request by stdpain**
## Proposed changes

use flag `-fno-strict-aliasing` won't generate SIMD Instructions for some vectorized case, so we need open it

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

### CODE REVIEW

File: be/src/util/binary_cast.hpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace doris {
+union TypeConverter {
+    uint64_t u64;
+    int64_t i64;
+    uint32_t u32[2];
+    int32_t i32[2];
+    uint8_t u8[8];
+    float flt[2];
+    double db;
```
Comment (yangzhg) [inline be/src/util/binary_cast.hpp:34]:
```suggestion
    double dbl;
```

**Review by HappenLee — DISMISSED** [review]
LGTM

**Comment by EmmyMiao87** [comment]
If we remove this flag '-fno-strict-aliasing', will forced conversion of different types of pointers produce unexpected results?

**Comment by stdpain** [comment]
@EmmyMiao87 
For example, an "unsigned int" can alias an "int", but not a "void*" or a "double".  A character type may alias any other type.

if we break this rules, gcc will generate a warning. I have fixed it.

**Comment by EmmyMiao87** [comment]
> @EmmyMiao87
> For example, an "unsigned int" can alias an "int", but not a "void*" or a "double". A character type may alias any other type.
> 
> if we break this rules, gcc will generate a warning. I have fixed it.

So if there is no similar warning in the compile stage, there will be no problems in the run stage, right?

**Comment by stdpain** [comment]
@EmmyMiao87 
right

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**